Home dind experiment code in test-infra

Q-Lee opened the pull request:
Home the dind test environment out of the k8s repo for now.

BenTheElder commented:
/assign
Looks like this just needs a fix to the boilerplate check and a `hack/update-bazel.sh`?

BenTheElder commented:
I think the makefile needs boilerplate now, possibly a new requirement? I remember @ixdy did something with the boilerplate presubmit recently.

ixdy commented on dind/Makefile:
```diff
@@ -0,0 +1,23 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+.PHONY: node cluster up
```
I think the boilerplate check wants a blank newline here. :man_shrugging:

BenTheElder replied:
yes

BenTheElder commented:
Could you add a README.md? We point a lot of people to docker-in-docker (IE our testing image with docker) in the context of ProwJobs and I'm concerned they will get lost down this rabbit hole :-)

Q-Lee commented:
I've added a README.

I grew tired of multi-minute cluster up, so I made two performance optimizations.

1. I build the base
I got tired of the super slow cluster up time, so I played with the file-system. I build the base with "--squash", but this requires docker experimental features, so I didn't do the same for the node and cluster images.

I also switched to the overlay driver. Theoretically, this makes our dind less portable. In practice, this driver is nearly ubiquitous. I didn't move to the more performant overlay2, because it's not yet available on COS.

Q-Lee commented:
I've added a README.

I grew tired of multi-minute cluster up, so I made two performance optimizations.

1. The base image has a lot of layers, so I squashed them. I didn't do the same for node or cluster, because it requires experimental features, and is a little CPU heavy.
1. I changed dind to use overlay. Theoretically, this is less portable than vfs. In practice, overlay's nearly ubiquitous. I didn't move to overlay2, because it's wasn't on our GKE COS nodes.

| Build/Driver | Time to up control plane (m:ss) |
| --- | --- |
| overlay/squash base | 0:42 |
| overlay2/no squashing | 0:41 |
| overlay2/squash base | 0:39 |
| overlay2/squash all images | 0:38 |
| vfs/no squashing | 4:25 |
| vfs/squash base | 2:10 |
| vfs/ squash all | 2:16 |

Squashing the base image always gives a performance bump, but it's tiny with overlay2. The biggest improvement is changing the driver. I'm still going to squash the base, because it will make things a lot better if we *do* ever run on vfs.

BenTheElder commented:
Nice.

... Did you mean to close this?

Q-Lee commented:
No, I did not mean to click that button.

BenTheElder commented:
probably just `/hold cancel` this in the morning @Q-Lee

Q-Lee commented:
/hold cancel
